lib/transbank/sdk/onepay/models/item.rb

```diff
@@ -22,74 +19,86 @@ class Item
+        self.expire = opts.fetch(:expire, nil)
       end
 
       # @param description [String] An item's description
-      # @raise [ItemError] when description is not [String]
+      # @raise [ItemError] when description is null
       def description=(description)
-        unless description.is_a? String
-          raise ItemError "Description is not a String"
-        end
+        raise Errors::ItemError, "Description cannot be null" if description.nil?
         @description = description
       end
 
       # @param quantity [Integer] How many of units of [Item]
-      # @raise [ItemError] when given quantity is not an [Integer] or is less than zero
+      # @raise [ItemError] when given quantity is nil or less than zero.
       def quantity=(quantity)
-        unless quantity.is_a? Integer
-          raise ItemError "Quantity must be an Integer"
-        end
-
+        raise Errors::ItemError, "Quantity cannot be null" if quantity.nil?
         if quantity < 0
-          raise ItemError "Quantity cannot be less than zero"
+          raise Errors::ItemError, "Quantity cannot be less than zero"
         end
         @quantity = quantity
       end
 
       # @param amount [Integer] The value of each unit of [Item]
-      # @raise [ItemError] when amount is not an [Integer] or is less than zero.
+      # @raise [ItemError] when amount is nil or less than zero.
       def amount=(amount)
-        unless amount.is_a? Integer
-          raise ItemError "Amount must be an Integer"
-        end
-
+        raise Errors::ItemError, "Amount cannot be null" if amount.nil?
         if amount < 0
-          raise ItemError "Amount cannot be less than zero"
+          raise Errors::ItemError, "Amount cannot be less than zero"
         end
         @amount = amount
       end
 
       # @param additional_data [String] A string with whatever additional data the
       # Merchant might want to add
-      # @raise [ItemError] when additional_data is not a [String]
       def additional_data=(additional_data)
         additional_data = '' if additional_data.nil?
-        unless additional_data.is_a? String
-          raise ItemError 'Additional data must be a String'
-        end
         @additional_data = additional_data
       end
 
       # @param expire [Integer] Expiry for the Item
-      # @raise [ItemError] if value is not an [Integer]
       def expire=(expire)
-        unless expire.is_a? Integer
-          raise ItemError 'Expire must be an Integer'
-        end
+        expire = expire || 0
         @expire = expire
       end
+
+      # Return the total amount to pay for the Item, that is, amount * quantity
+      # @return [Numeric] the total amount to pay for the [Item]
+      def total
+        self.quantity * self.amount
+      end
+
+      # Override == to allow comparison between [Item]s
+      # @return [boolean] true if equal, false otherwise
+      def ==(another_item)
+        instance_variables.map! { |var| var.to_s.gsub!(/^@/, '') }
```

Comment:
This is very low on my readability/mantainability score. You sure you are getting a lot of value for this hard to read code?. I love metaprogramming, but you have to be careful where you use it.

Reply:
I'd say you better write manually the comparison of all the fields explicitely, if you *really* need this method to work. BTW, why do you need it?

Reply:
This is done so the #remove method on ShoppingCart works correctly (without having to write code similar to this or your proposed solution on the #remove method/ShoppingCart itself)

Reply:
Agree on low readabilty tho

Reply:
Then if you really need it, implement this method in the boring, explicit way